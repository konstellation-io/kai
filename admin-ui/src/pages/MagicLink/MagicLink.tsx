import { get } from 'lodash';

import React, { useEffect, useState, useRef } from 'react';
import useEndpoint from '../../hooks/useEndpoint';
import { useParams, useHistory } from 'react-router-dom';

import { ENDPOINT } from '../../constants/application';
import ROUTE from '../../constants/routes';
import * as CHECK from '../../components/Form/check';
import Circle from '../../components/Shape/Circle/Circle';
import { STATES } from '../../constants/application';

import styles from './MagicLink.module.scss';

const DEFAULT_ERROR = 'Unexpected error. Contact support for more information';

const getTitle = {
  [STATES.INITIALIZING]: 'Connecting to KRE',
  [STATES.SUCCESS]: 'You are connected!',
  [STATES.ERROR]: 'Cannot connect to KRE'
};
const getSubtitle = {
  [STATES.INITIALIZING]: 'It will take a moment',
  [STATES.SUCCESS]:
    'Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. ',
  [STATES.ERROR]:
    'Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. '
};
const getCircleText = {
  [STATES.INITIALIZING]: 'LOADING',
  [STATES.SUCCESS]: 'DONE',
  [STATES.ERROR]: 'ERROR'
};

function checkToken(token: string) {
  return CHECK.getValidationError([CHECK.isFieldNotEmpty(token)]);
}

interface RouteParams {
  token: string;
}

/**
 * Design requirements:
 *  - Loading animation must be shown for a short period of time
 *  - After loading animation time, if there is a response or an error the
 *    status of the page must be updated.
 *  - The SUCCESS animation must be shown for another period of time before leaving
 *    the view.
 */
function MagicLink() {
  const history = useHistory();
  const [error, setError] = useState<string | boolean>('');
  const [animationPending, setAnimationPending] = useState(true);
  const [status, setStatus] = useState(STATES.INITIALIZING);
  const [response, makeRequest] = useEndpoint({
    endpoint: ENDPOINT.VALIDATE_MAGIC_LINK,
    method: 'POST'
  });
  const timeout = useRef<number>();
  const { token } = useParams<RouteParams>();

  // Checks for token errors and send login request
  useEffect(function() {
    const err = checkToken(token);

    if (err === true) {
      makeRequest({ verificationCode: token });
    } else {
      setError(err);
    }

    timeout.current = window.setTimeout(() => {
      setAnimationPending(false);
    }, 3000);

    return function() {
      if (timeout.current) window.clearTimeout(timeout.current);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Updates status after getting a response and waiting for loading animation.
  useEffect(
    function() {
      if (!animationPending) {
        timeout.current = undefined;
        if (error) {
          setStatus(STATES.ERROR);
          setError(error || DEFAULT_ERROR);
        } else if (response.complete) {
          if (response.status === 200) {
            setStatus(STATES.SUCCESS);

            timeout.current = window.setTimeout(() => {
              history.push(ROUTE.HOME);
            }, 2500);
          } else if (
            response.error &&
            get(response, 'data.code') === 'invalid_verification_code'
          ) {
            setError('Invalid verification code');
          } else {
            setError('Unexpected error. Contact support for more information');
          }
        }
      }
    },
    [response, history, animationPending, status, error]
  );

  return (
    <div className={styles.bg}>
      <div className={styles.grid}>
        <div className={styles.container}>
          <h1>{getTitle[status]}</h1>
          <p className={styles.subtitle}>{getSubtitle[status]}</p>
          <Circle animation={status} label={getCircleText[status]} />
          {status === STATES.ERROR && (
            <div className={styles.errorText}>{error}</div>
          )}
        </div>
      </div>
    </div>
  );
}

export default MagicLink;
