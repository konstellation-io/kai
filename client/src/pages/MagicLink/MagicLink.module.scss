@import '../../styles/variables';
@import '../../styles/colors';

.bg {
  width: 100%;
  height: 100%;
  background-color: black;
}

.grid {
  display: grid;
  grid-template-columns: repeat(12, minmax(56px, 1fr));
  grid-gap: 24px;
  width: 100%;
  height: 100%;
  max-width: $container-max-width-login;
  margin: 0 auto;
}

.container {
  grid-area: 1 / 3 / 2 / 11;
  height: 100%;
  padding: $form-padding-height 0;
  background-color: $bg-color;
}

.error {
  position: absolute;
  top: 25%;
  left: 50%;
  transform: translate(-50%, -50%);
  color: $color-red;
}
