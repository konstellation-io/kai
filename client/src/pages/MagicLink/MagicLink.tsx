import React, { useEffect, useState } from 'react';
import useEndpoint from '../../hooks/useEndpoint';
import { parse } from 'query-string';

import Spinner from '../../components/Spinner/Spinner';
import * as CHECK from '../../components/Form/check';
import { ENDPOINT } from '../../constants/application';
import * as PAGES from '../../constants/routes';

import styles from './MagicLink.module.scss';


type Props = {
  location: any;
  history: any;
};

export function getToken(location:Location) {
  // @ts-ignore
  return parse(location.search, { ignoreQueryPrefix: true }).token || '';
}

function checkToken(token:string) {
  return CHECK.getValidationError([
    CHECK.isFieldNotEmpty(token),
    CHECK.isMagicLinkTokenValid(token)
  ]);
}

function MagicLink({ location, history }: Props) {
  const [error, setError] = useState('');
  const [response, makeRequest] = useEndpoint({
    endpoint: ENDPOINT.VALIDATE_MAGIC_LINK,
    method: 'POST'
  });

  const token = getToken(location);

  useEffect(function() {
    const error = checkToken(token);
    
    if (!error) {
      makeRequest({ token });
    } else {
      setError(error);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(function() {
    if (response.complete) {
      if (response.data === 'OK') {
        history.push(PAGES.HOME);
      } else {
        setError('Unexpected error. Contact support for more information');
      }
    }
  }, [response, history]);

  return (
    <div className={ styles.bg }>
      <div className={ styles.grid }>
        <div className={ styles.container }>
          { response.pending
              ? <Spinner size={ 60 } />
              : <p className={ styles.error }>{ error }</p>
          }
        </div>
      </div>
    </div>
  );
}

export default MagicLink;
