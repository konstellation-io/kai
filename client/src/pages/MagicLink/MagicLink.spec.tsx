import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, createLocation } from 'history';
import { render, cleanup, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'
import MagicLink, { getToken } from './MagicLink';
import axios from 'axios';

jest.mock('axios');

afterEach(cleanup);


it('Render MagicLink without crashing', () => {
  const history = createMemoryHistory();
  const location = createLocation('/magic_link');
  const { container } = render(
    <Router history={history}>
      <MagicLink history={history} location={ location } />
    </Router>
  );
  expect(container).toMatchSnapshot();
});

it('obtains token from location', () => {
  const location:any = createLocation('/magic_link?token=123456');
  const token = getToken(location);

  expect(token).toBe('123456');
});

it('handles success response', async () => {
  // @ts-ignore
  axios.mockResolvedValue({ data: 'OK' });

  const history = createMemoryHistory();
  const location = createLocation('/magic_link?token=123456');

  await act(async () => {
    await render(
      <Router history={history}>
        <MagicLink history={history} location={ location } />
      </Router>
    );
  });
});

it('handles error response', async () => {
  // @ts-ignore
  axios.mockRejectedValue({ data: 'ERROR' });

  let getByText = function(selector:string) {};
  const history = createMemoryHistory();
  const location = createLocation('/magic_link?token=123456');

  await act(async () => {
    ({ getByText } = await render(
      <Router history={history}>
        <MagicLink history={history} location={ location } />
      </Router>
    ));
  });

  expect(getByText('Unexpected error. Contact support for more information')).toBeInTheDocument();
});
