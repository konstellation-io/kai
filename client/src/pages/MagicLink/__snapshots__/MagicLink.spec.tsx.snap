// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MagicLink without crashing 1`] = `
<div>
  <div
    class="bg"
  >
    <div
      class="grid"
    >
      <div
        class="container"
      >
        <p
          class="error"
        >
          This field cannot be empty
        </p>
      </div>
    </div>
  </div>
</div>
`;
